src/java_bytecode/synthetic_methods_map.h

```diff
@@ -9,12 +9,30 @@ Author: Chris Smowton, chris.smowton@diffblue.com
 #ifndef CPROVER_JAVA_BYTECODE_SYNTHETIC_METHODS_MAP_H
 #define CPROVER_JAVA_BYTECODE_SYNTHETIC_METHODS_MAP_H
 
+/// \file
+///   Synthetic methods are particular methods internally generated by the
+///   Java frontend (cf. normal methods which are translated from Java
+///   bytecode). This file provides an enumeration specifying the kind of a
```

Comment:
Are these otherwise known as intrinsics? Or built-in functions? I'm still of the opinion that new terminology has been invented here.

Reply:
I don't think built in is quite right since they are generated on the fly and there are a bunch of them (one clinit_wrapper per input class). I'm not sure I really know what would be meant by an intrinsic method but if there is prior usage of this then maybe it covers it?

Reply:
Wikipedia says: https://en.wikipedia.org/wiki/Intrinsic_function

Reply:
I wouldn't say intrinsic-- that would be something more like a primitive operation that's called like a function, like a compare-and-set operation for example. These are GOTO methods without corresponding user code. __CPROVER_start is also like that, what do you call that?

Reply:
Are these [thunks](https://en.wikipedia.org/wiki/Thunk)? Maybe you can give me some examples of "synthetic methods" in the Java front-end? The comment that has been put in place only gives examples of methods that are _not_ considered synthetic methods.

Reply:
Thunks seems accurate IMO.

Reply:
The examples are in this very file, see the enum `synthetic_method_typet`

One kind of synthetic method (`STATIC_INITIALIZER_WRAPPER`) is a thunk (it wraps a callee with an if-already-run test, which is a pretty classic call-by-need thunk).

The other kind (`STUB_CLASS_STATIC_INITIALIZER`) isn't really, it's a class static initialiser, except that whereas normal ones are generated from user bytecode, these ones are made up ("synthesised"? :)) by the frontend to initialise the fields of an opaque type.

Reply:
If you really need that umbrella term then so be it. It just seems to me that all the steps happening here are pretty standard for compilers and loaders. These tools regularly add code that isn't just a translation from user input, but instead is necessary for execution on a particular platform or is mandated by the language specification. Some of them are default constructors (these are specific to the types used in the input program), some others are built-in functions (those have fixed code baked into the compiler and will be added on an as-needed basis), others are thunks. So overall we don't seem to be doing anything non-standard here, and hence I'm unhappy that we come up with a non-standard term. But if you do need a new term to describe the combination of all these standard steps then you're probably just doing the right thing.

Reply:
Basically I need a name for the abstract supertype of those things, so when an outsider asks "make the body for method X!" I can dispatch on the particular kind of method I'm supposed to be fleshing out.

Also in my defence "synthesized" seems to be a pretty common term for compiler-generated default constructors, for example, and Java bytecode also uses "synthetic" for methods and fields that don't exist in the user's input code, but are created by the compiler.